### PR TITLE
Introduce private environment variables for overriding http retries

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/11590?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11590/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11590
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by jameswu-db — APPROVED** [review]
thank you for the quick turnaround!

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

File: mlflow/utils/rest_utils.py [hunk 1]
```diff
@@ -183,6 +186,41 @@ def verify_rest_response(response, endpoint):
...
+            "Cannot be negative."
+        )
+    if max_retries >= max_retry_limit:
+        raise MlflowException(
+            f"The configured max_retries ({max_retries}) are "
+            f"in excess of the maximum allowable retries ({max_retry_limit})"
+        )
+    if backoff_factor >= max_backoff_factor_limit:
+        raise MlflowException(
+            f"The configured backoff_factor ({backoff_factor}) are in excess "
```
Comment (harupy) [inline mlflow/utils/rest_utils.py:210]:
```suggestion
            f"The configured backoff_factor ({backoff_factor}) is in excess "
```

should we use `is` here?

File: mlflow/utils/rest_utils.py [hunk 2]
```diff
@@ -183,6 +186,41 @@ def verify_rest_response(response, endpoint):
     return response
 
 
+def _validate_retries_and_backoff_values(max_retries, backoff_factor):
```
Comment (harupy) [inline mlflow/utils/rest_utils.py:189]: Can we split this?

File: mlflow/utils/rest_utils.py [hunk 3]
```diff
@@ -183,6 +186,41 @@ def verify_rest_response(response, endpoint):
...
+    if max_retries < 0:
+        raise MlflowException(
+            f"The max_retries value must be either 0 a positive integer. Got {max_retries}"
+        )
+    if backoff_factor < 0:
+        raise MlflowException(
+            f"The backoff_factor value must be either 0 a positive integer. Got {backoff_factor}"
+        )
+
+
```
Comment (harupy) [inline mlflow/utils/rest_utils.py:223]: /ai split this function into two functions, one for `max_retries` and one for `backoff_factor`.

Comment (harupy) [reply mlflow/utils/rest_utils.py:223]: @BenWilson2 I think this is easier to follow.

File: mlflow/utils/rest_utils.py [hunk 4]
```diff
@@ -183,6 +186,41 @@ def verify_rest_response(response, endpoint):
...
+            f"({max_backoff_factor_limit})"
+        )
+    if max_retries < 0:
+        raise MlflowException(
+            f"The max_retries value must be either 0 a positive integer. Got {max_retries}"
+        )
+    if backoff_factor < 0:
+        raise MlflowException(
+            f"The backoff_factor value must be either 0 a positive integer. Got {backoff_factor}"
+        )
```
Comment (harupy) [inline mlflow/utils/rest_utils.py:221]:
```suggestion
        raise MlflowException(
            f"The backoff_factor value must be either 0 a positive integer. Got {backoff_factor}",
            invalid_parameter_value
        )
```

File: mlflow/utils/rest_utils.py (hunk 3 above)
Comment (harupy) [reply mlflow/utils/rest_utils.py:223]: We don't need `_values` suffix.

**Review by harupy — APPROVED** [review]
LGTM once comments are addressed!

File: tests/utils/test_rest_utils.py [hunk 5]
```diff
@@ -602,3 +602,45 @@ def test_provide_redirect_kwarg():
...
+def test_http_request_max_retries(monkeypatch):
+    monkeypatch.setenv("_MLFLOW_HTTP_REQUEST_MAX_RETRIES_LIMIT", "15")
+    host_creds = MlflowHostCreds("http://example.com")
+
+    with mock.patch("requests.Session.request") as mock_request:
+        with pytest.raises(MlflowException, match="The configured max_retries"):
+            http_request(host_creds, "/endpoint", "GET", max_retries=16)
+        mock_request.assert_not_called()
+        http_request(host_creds, "/endpoint", "GET", max_retries=3)
+        mock_request.assert_called()
```
Comment (harupy) [inline tests/utils/test_rest_utils.py:616]:
```suggestion
        mock_request.assert_called_once()
```

nit: for more strict validation.

File: mlflow/utils/rest_utils.py (hunk 3 above)
Comment (BenWilson2) [reply mlflow/utils/rest_utils.py:223]: Good idea! will split to reduce the overloading of responsibility

File: mlflow/utils/rest_utils.py (hunk 4 above)
Comment (BenWilson2) [reply mlflow/utils/rest_utils.py:221]: Ah, good catch!

**Pull request merged**